// @zegocloud/zego-uikit-rn/lib/commonjs/signal_plugin/defines.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZegoUIKitPluginType = exports.ZIMConnectionState = exports.ZIMCallUserState = exports.CXCallEndedReason = void 0;
const ZegoUIKitPluginType = exports.ZegoUIKitPluginType = {
  signaling: 1
};
let CXCallEndedReason = exports.CXCallEndedReason = /*#__PURE__*/function (CXCallEndedReason) {
  CXCallEndedReason[CXCallEndedReason["Failed"] = 1] = "Failed";
  CXCallEndedReason[CXCallEndedReason["RemoteEnded"] = 2] = "RemoteEnded";
  CXCallEndedReason[CXCallEndedReason["Unanswered"] = 3] = "Unanswered";
  CXCallEndedReason[CXCallEndedReason["AnsweredElsewhere"] = 4] = "AnsweredElsewhere";
  CXCallEndedReason[CXCallEndedReason["DeclinedElsewhere"] = 5] = "DeclinedElsewhere";
  return CXCallEndedReason;
}({});
let ZIMConnectionState = exports.ZIMConnectionState = /*#__PURE__*/function (ZIMConnectionState) {
  ZIMConnectionState[ZIMConnectionState["Disconnected"] = 0] = "Disconnected";
  ZIMConnectionState[ZIMConnectionState["Connecting"] = 1] = "Connecting";
  ZIMConnectionState[ZIMConnectionState["Connected"] = 2] = "Connected";
  ZIMConnectionState[ZIMConnectionState["Reconnecting"] = 3] = "Reconnecting";
  return ZIMConnectionState;
}({});
let ZIMCallUserState = exports.ZIMCallUserState = /*#__PURE__*/function (ZIMCallUserState) {
  ZIMCallUserState[ZIMCallUserState["Inviting"] = 0] = "Inviting";
  ZIMCallUserState[ZIMCallUserState["Accepted"] = 1] = "Accepted";
  ZIMCallUserState[ZIMCallUserState["Rejected"] = 2] = "Rejected";
  ZIMCallUserState[ZIMCallUserState["Cancelled"] = 3] = "Cancelled";
  ZIMCallUserState[ZIMCallUserState["Offline"] = 4] = "Offline";
  ZIMCallUserState[ZIMCallUserState["Received"] = 5] = "Received";
  return ZIMCallUserState;
}({});
