// @zegocloud/zego-uikit-rn/lib/module/signal_plugin/defines.js
export const ZegoUIKitPluginType = {
  signaling: 1
};
export let CXCallEndedReason = /*#__PURE__*/function (CXCallEndedReason) {
  CXCallEndedReason[CXCallEndedReason["Failed"] = 1] = "Failed";
  CXCallEndedReason[CXCallEndedReason["RemoteEnded"] = 2] = "RemoteEnded";
  CXCallEndedReason[CXCallEndedReason["Unanswered"] = 3] = "Unanswered";
  CXCallEndedReason[CXCallEndedReason["AnsweredElsewhere"] = 4] = "AnsweredElsewhere";
  CXCallEndedReason[CXCallEndedReason["DeclinedElsewhere"] = 5] = "DeclinedElsewhere";
  return CXCallEndedReason;
}({});
export let ZIMConnectionState = /*#__PURE__*/function (ZIMConnectionState) {
  ZIMConnectionState[ZIMConnectionState["Disconnected"] = 0] = "Disconnected";
  ZIMConnectionState[ZIMConnectionState["Connecting"] = 1] = "Connecting";
  ZIMConnectionState[ZIMConnectionState["Connected"] = 2] = "Connected";
  ZIMConnectionState[ZIMConnectionState["Reconnecting"] = 3] = "Reconnecting";
  return ZIMConnectionState;
}({});
export let ZIMCallUserState = /*#__PURE__*/function (ZIMCallUserState) {
  ZIMCallUserState[ZIMCallUserState["Inviting"] = 0] = "Inviting";
  ZIMCallUserState[ZIMCallUserState["Accepted"] = 1] = "Accepted";
  ZIMCallUserState[ZIMCallUserState["Rejected"] = 2] = "Rejected";
  ZIMCallUserState[ZIMCallUserState["Cancelled"] = 3] = "Cancelled";
  ZIMCallUserState[ZIMCallUserState["Offline"] = 4] = "Offline";
  ZIMCallUserState[ZIMCallUserState["Received"] = 5] = "Received";
  return ZIMCallUserState;
}({});
