// @zegocloud/zego-uikit-rn/lib/module/components/audio_video/ZegoLeaveButton.js
import React from "react";
import { Image, TouchableOpacity } from "react-native";
import ZegoUIKitInternal from "../internal/ZegoUIKitInternal";
export default function ZegoLeaveButton(props) {
  const {
    iconLeave,
    onLeaveConfirmation,
    onPressed,
    width = 48,
    height = 48
  } = props;
  const onPress = () => {
    if (typeof onLeaveConfirmation == 'function') {
      onLeaveConfirmation().then(() => {
        ZegoUIKitInternal.leaveRoom();
        if (typeof onPressed == 'function') {
          onPressed();
        }
      });
    } else {
      ZegoUIKitInternal.leaveRoom();
      if (typeof onPressed == 'function') {
        onPressed();
      }
    }
  };
  return /*#__PURE__*/React.createElement(TouchableOpacity, {
    style: {
      width: width,
      height: height
    },
    onPress: onPress
  }, /*#__PURE__*/React.createElement(Image, {
    resizeMode: "contain",
    source: iconLeave ? iconLeave : require("../internal/resources/white_button_hang_up.png"),
    style: {
      width: "100%",
      height: "100%"
    }
  }));
}
