// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/services/defines.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZegoTranslationText = exports.ZegoToastType = exports.ZegoLiveStreamingRole = exports.ZegoLiveStatus = exports.ZegoInvitationType = exports.ZegoCoHostConnectState = exports.HOST_DEFAULT_CONFIG = exports.AUDIENCE_DEFAULT_CONFIG = void 0;
var _ZegoMenuBarButtonName = _interopRequireDefault(require("../components/ZegoMenuBarButtonName"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
class ZegoDialogInfo {
  constructor(_ref) {
    let {
      title,
      message,
      cancelButtonName,
      confirmButtonName
    } = _ref;
    _defineProperty(this, "title", '');
    _defineProperty(this, "message", '');
    _defineProperty(this, "cancelButtonName", '');
    _defineProperty(this, "confirmButtonName", '');
    this.title = title;
    this.message = message;
    this.cancelButtonName = cancelButtonName || 'Cancel';
    this.confirmButtonName = confirmButtonName || 'OK';
  }
}
;
const ZegoTranslationText = {
  startLiveStreamingButton: "Start",
  endCoHostButton: "End",
  requestCoHostButton: "Apply to co-host",
  cancelRequestCoHostButton: "Cancel the application",
  removeCoHostButton: "Remove the co-host",
  inviteCoHostButton: "Invite %0 to co-host",
  cancelMenuDialogButton: "Cancel",
  noHostOnline: "No host is online.",
  memberListTitle: "Audience",
  sendRequestCoHostToast: "You are applying to be a co-host, please wait for confirmation.",
  // green
  hostRejectCoHostRequestToast: "Your request to co-host with the host has been refused.",
  // red
  inviteCoHostFailedToast: "Failed to connect with the co-host, please try again.",
  // red
  repeatInviteCoHostFailedToast: "You've sent the co-host invitation, please wait for confirmation.",
  // red
  audienceRejectInvitationToast: "%0 refused to be a co-host.",
  // red
  requestCoHostFailed: "Failed to apply for connection.",
  // red
  receivedRequestCoHostToast: "%0 wants to co-host with you.",
  // green
  inviteCoHostSuccessfullyToast: "Sent the invitation successfully.",
  // green
  removeUserMenuDialogButton: "Remove %0 from the room",
  cameraPermissionSettingDialogInfo: new ZegoDialogInfo({
    title: "Can not use Camera!",
    message: "Please enable camera access in the system settings!",
    cancelButtonName: "Cancel",
    confirmButtonName: "Settings"
  }),
  microphonePermissionSettingDialogInfo: new ZegoDialogInfo({
    title: "Can not use Microphone!",
    message: "Please enable microphone access in the system settings!",
    cancelButtonName: "Cancel",
    confirmButtonName: "Settings"
  }),
  receivedCoHostInvitationDialogInfo: new ZegoDialogInfo({
    title: "Invitation",
    message: "The host is inviting you to co-host.",
    cancelButtonName: "Disagree",
    confirmButtonName: "Agree"
  }),
  endConnectionDialogInfo: new ZegoDialogInfo({
    title: "End the connection",
    message: "Do you want to end the cohosting?",
    cancelButtonName: 'Cancel',
    confirmButtonName: 'OK'
  })
};
exports.ZegoTranslationText = ZegoTranslationText;
const ZegoLiveStreamingRole = {
  host: 0,
  coHost: 1,
  audience: 2
};
exports.ZegoLiveStreamingRole = ZegoLiveStreamingRole;
const ZegoLiveStatus = {
  default: 0,
  start: 1
};
exports.ZegoLiveStatus = ZegoLiveStatus;
const ZegoInvitationType = {
  requestCoHost: 2,
  inviteToCoHost: 3,
  removeCoHost: 4
};
exports.ZegoInvitationType = ZegoInvitationType;
const ZegoCoHostConnectState = {
  idle: 0,
  connecting: 1,
  connected: 2
};
exports.ZegoCoHostConnectState = ZegoCoHostConnectState;
const ZegoToastType = {
  default: 0,
  info: 1,
  success: 2,
  warning: 3,
  error: 4
};
exports.ZegoToastType = ZegoToastType;
const HOST_DEFAULT_CONFIG = {
  role: ZegoLiveStreamingRole.host,
  turnOnCameraWhenJoining: true,
  turnOnMicrophoneWhenJoining: true,
  useSpeakerWhenJoining: true,
  audioVideoViewConfig: {
    showSoundWavesInAudioMode: true
  },
  bottomMenuBarConfig: {
    showInRoomMessageButton: true,
    hostButtons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchCameraButton],
    coHostButtons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchCameraButton, _ZegoMenuBarButtonName.default.coHostControlButton],
    audienceButtons: [_ZegoMenuBarButtonName.default.coHostControlButton],
    hostExtendButtons: [],
    coHostExtendButtons: [],
    audienceExtendButtons: [],
    maxCount: 5
  },
  confirmDialogInfo: new ZegoDialogInfo({
    title: "Stop the live",
    message: "Are you sure to stop the live?",
    cancelButtonName: "Cancel",
    confirmButtonName: "Stop it"
  }),
  translationText: ZegoTranslationText
};
exports.HOST_DEFAULT_CONFIG = HOST_DEFAULT_CONFIG;
const AUDIENCE_DEFAULT_CONFIG = {
  role: ZegoLiveStreamingRole.audience,
  turnOnCameraWhenJoining: false,
  turnOnMicrophoneWhenJoining: false,
  useSpeakerWhenJoining: true,
  audioVideoViewConfig: {
    showSoundWavesInAudioMode: true
  },
  bottomMenuBarConfig: {
    showInRoomMessageButton: true,
    hostButtons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchCameraButton],
    coHostButtons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchCameraButton, _ZegoMenuBarButtonName.default.coHostControlButton],
    audienceButtons: [_ZegoMenuBarButtonName.default.coHostControlButton],
    hostExtendButtons: [],
    coHostExtendButtons: [],
    audienceExtendButtons: [],
    maxCount: 5
  },
  translationText: ZegoTranslationText
};
exports.AUDIENCE_DEFAULT_CONFIG = AUDIENCE_DEFAULT_CONFIG;
