// @zegocloud/zego-uikit-rn/lib/commonjs/signal_plugin/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _core = _interopRequireDefault(require("../core"));
var _defines = _interopRequireDefault(require("../core/defines"));
var _logger = require("../utils/logger");
var _ZegoUIKitCorePlugin = _interopRequireDefault(require("../../components/internal/ZegoUIKitCorePlugin"));
var _reactNative = require("react-native");
var _class;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
class ZegoPluginInvitationService {
  constructor() {
    _defineProperty(this, "_notifyWhenAppRunningInBackgroundOrQuit", void 0);
    _defineProperty(this, "_androidOfflineDataHandler", void 0);
    _defineProperty(this, "_iOSOfflineDataHandler", void 0);
    _defineProperty(this, "_callKitAnswerCallHandler", void 0);
    _defineProperty(this, "_callKitEndCallHandler", void 0);
    if (!ZegoPluginInvitationService.shared) {
      this._notifyWhenAppRunningInBackgroundOrQuit = true;
      ZegoPluginInvitationService.shared = this;
    }
    return ZegoPluginInvitationService.shared;
  }
  static getInstance() {
    if (!ZegoPluginInvitationService.shared) {
      ZegoPluginInvitationService.shared = new ZegoPluginInvitationService();
    }
    return ZegoPluginInvitationService.shared;
  }
  setBackgroundMessageHandler() {
    _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.setBackgroundMessageHandler(message => {
      (0, _logger.zloginfo)('ZPNs setBackgroundMessageHandler: ', message);
      const dataObj = JSON.parse(message.extras.payload);
      dataObj.zim_call_id = message.extras.call_id;
      ZegoPluginInvitationService.getInstance().getAndroidOfflineDataHandler()(dataObj);
    });
  }
  setAndroidOfflineDataHandler(handler) {
    this._androidOfflineDataHandler = handler;
  }
  getAndroidOfflineDataHandler() {
    return this._androidOfflineDataHandler;
  }
  setIOSOfflineDataHandler(handler) {
    this._iOSOfflineDataHandler = handler;
  }
  onCallKitAnswerCall(handler) {
    this._callKitAnswerCallHandler = handler;
  }
  onCallKitEndCall(handler) {
    this._callKitEndCallHandler = handler;
  }
  getIOSOfflineDataHandler() {
    return this._iOSOfflineDataHandler;
  }
  getAnswerCallHandle() {
    return this._callKitAnswerCallHandler;
  }
  getEndCallHandle() {
    return this._callKitEndCallHandler;
  }
  reportCallKitCallEnded(uuid, reason) {
    return _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().reportCallEnded(reason, uuid);
  }
  getZIMInstance() {
    return _core.default.getInstance().getZIMInstance();
  }
  getVersion() {
    return _core.default.getInstance().getVersion();
  }
  init(appID, appSign) {
    _core.default.getInstance().create({
      appID,
      appSign
    });
  }
  uninit() {
    _core.default.getInstance().destroy();
  }
  login(userID, userName, token) {
    return _core.default.getInstance().login({
      userID,
      userName
    }, token);
  }
  logout() {
    return _core.default.getInstance().logout();
  }
  enableNotifyWhenAppRunningInBackgroundOrQuit(enable, isIOSDevelopmentEnvironment, appName) {
    this._notifyWhenAppRunningInBackgroundOrQuit = enable;
    if (enable) {
      if (_reactNative.Platform.OS === 'ios') {
        const CXProviderConfiguration = {
          localizedName: appName ?? 'My app',
          iconTemplateImageName: "AppIcon"
        };
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.setInitConfiguration(CXProviderConfiguration);
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().applyNotificationPermission();
        const iOSEnvironment = isIOSDevelopmentEnvironment == null ? 2 : isIOSDevelopmentEnvironment ? 1 : 0;
        console.log('#########registerPush, iOSEnvironment', iOSEnvironment);
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().registerPush({
          enableIOSVoIP: true,
          iOSEnvironment: iOSEnvironment
        });
      } else {
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.setPushConfig({
          "enableFCMPush": true,
          "enableHWPush": false,
          "enableMiPush": false,
          "enableOppoPush": false,
          "enableVivoPush": false
        });
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().registerPush({
          enableIOSVoIP: true
        });
      }
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().on("registered", message => {
        console.log("@@@@@@@@@@@@@@@@>>>>>>>>>>>>>>>############", message);
      });

      // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().on("notificationArrived", (message) => {
      //   console.log("@@@@@@@@@@@@@@@@notificationArrived>>>>>>>>>>>>>>>############", getCallID(message))
      //   setZpnState("notificationArrived: " + getCallID(message))
      // })
      // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().on("notificationClicked", (message) => {
      //   console.log("@@@@@@@@@@@@@@@@notificationClicked>>>>>>>>>>>>>>>############", getCallID(message))
      //   setZpnState("notificationClicked: " + getCallID(message))
      // })
      // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().on("throughMessageReceived", (message) => {
      //   console.log("@@@@@@@@@@@@@@@@throughMessageReceived>>>>>>>>>>>>>>>############", getCallID(message))
      //   setZpnState("throughMessageReceived: " + getCallID(message))
      // })

      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("didReceiveIncomingPush", (extras, uuid) => {
        console.log('#########didReceiveIncomingPush', extras, uuid);
        let {
          payload
        } = extras;
        let {
          call_id
        } = extras;
        const dataObj = payload ? JSON.parse(payload) : {};
        dataObj.zim_call_id = call_id;
        ZegoPluginInvitationService.getInstance().getIOSOfflineDataHandler()(dataObj, uuid);
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("providerDidReset", () => {
        console.log('#########providerDidReset');
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("providerDidBegin", () => {
        console.log('#########providerDidBegin');
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("didActivateAudioSession", () => {
        console.log('#########didActivateAudioSession');
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("didDeactivateAudioSession", () => {
        console.log('#########didDeactivateAudioSession');
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("timedOutPerformingAction", action => {
        console.log('#########timedOutPerformingAction', action);
        action.fulfill();
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performStartCallAction", action => {
        console.log('#########performStartCallAction', action);
        action.fulfill();
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performAnswerCallAction", action => {
        console.log('#########performAnswerCallAction', action);
        ZegoPluginInvitationService.getInstance().getAnswerCallHandle()(action);
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performEndCallAction", action => {
        console.log('#########performEndCallAction', action);
        ZegoPluginInvitationService.getInstance().getEndCallHandle()(action);
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performSetHeldCallAction", action => {
        console.log('#########performSetHeldCallAction', action);
        action.fulfill();
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performSetMutedCallAction", action => {
        console.log('#########performSetMutedCallAction', action);
        action.fulfill();
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performSetGroupCallAction", action => {
        console.log('#########performSetGroupCallAction', action);
        action.fulfill();
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().CallKit.getInstance().on("performPlayDTMFCallAction", action => {
        console.log('#########performPlayDTMFCallAction', action);
        action.fulfill();
      });
    } else {
      if (_ZegoUIKitCorePlugin.default.getZPNsPlugin()) {
        // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().unregisterPush();
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().off("registered");
        // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().off("notificationArrived")
        // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().off("notificationClicked")
        // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().off("throughMessageReceived")
      }
    }
  }

  sendInvitation(inviterName, invitees, timeout, type, data, notificationConfig) {
    // invitees = invitees.map((invitee) => invitee);
    if (!invitees.length) {
      (0, _logger.zlogerror)('[Service]Send invitees is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      timeout
    };
    config.extendedData = JSON.stringify({
      inviter_name: inviterName,
      type,
      data
    });
    if (this._notifyWhenAppRunningInBackgroundOrQuit) {
      config.pushConfig = {
        title: notificationConfig.title ?? "",
        content: notificationConfig.message ?? "",
        resourcesID: notificationConfig.resourceID ?? "",
        payload: data
      };
    }
    (0, _logger.zloginfo)(`[Service]Send invitation: invitees: ${invitees}, timeout: ${timeout}, type: ${type}, data: ${data}.`);
    return _core.default.getInstance().invite(invitees, config);
  }
  cancelInvitation(invitees, data, notificationConfig) {
    invitees = invitees.map(invitee => invitee);
    if (!invitees.length) {
      (0, _logger.zlogerror)('[Service]Cancel invitees is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      extendedData: data
    };
    if (this._notifyWhenAppRunningInBackgroundOrQuit) {
      config.pushConfig = {
        title: notificationConfig && notificationConfig.title,
        content: notificationConfig && notificationConfig.message,
        resourcesID: notificationConfig && notificationConfig.resourceID,
        payload: data
      };
    }
    const callID = _core.default.getInstance().getCallIDByUserID(_core.default.getInstance().getLocalUser().userID);
    (0, _logger.zloginfo)(`[Service]Cancel invitation: callID: ${callID}, invitees: ${invitees}, data: ${data}.`);
    return _core.default.getInstance().cancel(invitees, callID, config);
  }
  refuseInvitation(inviterID, data) {
    let callID;
    // Parse data and adapt automatic rejection
    if (data) {
      const dataObj = JSON.parse(data);
      callID = dataObj.callID;
    } else {
      callID = _core.default.getInstance().getCallIDByUserID(inviterID);
    }
    if (!callID) {
      (0, _logger.zlogerror)('[Service]Call id corresponding to the inviterID is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      extendedData: data
    };
    (0, _logger.zloginfo)(`[Service]Refuse invitation: callID: ${callID}, inviter id: ${inviterID}, data: ${data}.`);
    return _core.default.getInstance().reject(callID, config);
  }
  acceptInvitation(inviterID, data) {
    const callID = _core.default.getInstance().getCallIDByUserID(inviterID);
    if (!callID) {
      (0, _logger.zloginfo)('[Service]Call id corresponding to the inviterID is empty.', inviterID, data);
      return Promise.reject(new _defines.default());
    }
    const config = {
      extendedData: data
    };
    (0, _logger.zloginfo)(`[Service]Accept invitation: callID: ${callID}, inviter id: ${inviterID}, data: ${data}.`);
    return _core.default.getInstance().accept(callID, config);
  }
  queryCallList(count, nextFlag) {
    const config = {
      count: count,
      nextFlag: nextFlag
    };
    return _core.default.getInstance().queryCallList(config);
  }
  onConnectionStateChanged(callbackID, callback) {
    _core.default.getInstance().onConnectionStateChanged(callbackID, callback);
  }
  onCallInvitationReceived(callbackID, callback) {
    _core.default.getInstance().onCallInvitationReceived(callbackID, callback);
  }
  onCallInvitationTimeout(callbackID, callback) {
    _core.default.getInstance().onCallInvitationTimeout(callbackID, callback);
  }
  onCallInviteesAnsweredTimeout(callbackID, callback) {
    _core.default.getInstance().onCallInviteesAnsweredTimeout(callbackID, callback);
  }
  onCallInvitationAccepted(callbackID, callback) {
    _core.default.getInstance().onCallInvitationAccepted(callbackID, callback);
  }
  onCallInvitationRejected(callbackID, callback) {
    _core.default.getInstance().onCallInvitationRejected(callbackID, callback);
  }
  onCallInvitationCancelled(callbackID, callback) {
    _core.default.getInstance().onCallInvitationCancelled(callbackID, callback);
  }
}
exports.default = ZegoPluginInvitationService;
_class = ZegoPluginInvitationService;
_defineProperty(ZegoPluginInvitationService, "shared", void 0);
