// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoAudioVideoForegroundView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoAudioVideoForegroundView;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoAudioVideoForegroundView(props) {
  const {
    userInfo,
    showUserNameOnView,
    showCameraStateOnView,
    showMicrophoneStateOnView,
    turnOnMicrophoneWhenJoining,
    turnOnCameraWhenJoining,
    role,
    localUserID
  } = props;
  const {
    userID = '',
    userName = '',
    isMicDeviceOn
  } = userInfo;

  // const panResponder = useRef(PanResponder.create({
  //     onStartShouldSetPanResponderCapture: () => {
  //         console.log(`Toggle the state of other people's devices`);
  //     }
  // })).current;

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.foregroundViewContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.bottomContainer
  }, showUserNameOnView ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.nameLabelContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.nameLabel
  }, userName)) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), showMicrophoneStateOnView && !isMicDeviceOn ? /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoMicrophoneStateIcon, {
    userID: userID,
    style: styles.deviceIcon
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null)));
}
const styles = _reactNative.StyleSheet.create({
  foregroundViewContainer: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 10
  },
  topContainer: {
    paddingTop: 3,
    paddingRight: 3,
    paddingLeft: 3,
    paddingBottom: 3,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  bottomContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    backgroundColor: '#2A2A2A',
    opacity: 0.5,
    position: 'absolute',
    alignSelf: 'center',
    paddingLeft: 3,
    paddingRight: 3,
    paddingBottom: 1,
    paddingTop: 1,
    borderRadius: 4,
    bottom: 3,
    right: 3
  },
  nameLabelContainer: {
    alignSelf: 'center'
  },
  nameLabel: {
    color: '#FFFFFF',
    fontSize: 10
  },
  deviceIcon: {
    flex: 1,
    position: 'absolute'
  }
});
