// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoStartLiveButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoStartLiveButton;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _defines = require("../services/defines");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoStartLiveButton(props) {
  const {
    onPressed
  } = props;
  return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: onPressed,
    style: styles.btnContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.btnText
  }, _defines.ZegoTranslationText.startLiveStreamingButton));
}
const styles = _reactNative.StyleSheet.create({
  btnContainer: {
    width: 150,
    height: 44,
    borderRadius: 22,
    backgroundColor: 'rgba(167, 84, 255, 0.5)',
    alignItems: 'center',
    justifyContent: 'center'
  },
  btnText: {
    fontSize: 16,
    lineHeight: 22.5,
    color: '#fff'
  }
});
