// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoMinimizingButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoMinimizingButton;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _minimizing_helper = _interopRequireDefault(require("../services/minimizing_helper"));
var _utils = require("../utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoMinimizingButton(props) {
  const {
    width = 30,
    height = 30,
    borderRadius = 100,
    backgroundColor = 'rgba(30, 39, 64, 0.4000)',
    icon = require('../resources/icon_minimize.png'),
    onPressed,
    onWillPressed
  } = props;
  const pressedHandle = async () => {
    const canMinimize = await (0, _utils.getMethodReturnValue)(onWillPressed);
    if (canMinimize) {
      _minimizing_helper.default.getInstance().minimizeWindow();
      typeof onPressed === 'function' && onPressed();
    }
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    style: [styles.imgContainer, {
      width,
      height,
      backgroundColor,
      borderRadius
    }],
    onPress: pressedHandle
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: icon,
    style: {
      width: 18,
      height: 18
    }
  }));
}
const styles = _reactNative.StyleSheet.create({
  imgContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: 8
  }
});
