// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoCancelRequestCoHostButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoCancelRequestCoHostButton;
var _react = _interopRequireDefault(require("react"));
var _defines = require("../services/defines");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoCancelRequestCoHostButton(props) {
  const {
    hostID,
    onCancelSuccessfully,
    setIsToastVisable,
    setToastExtendedData
  } = props;
  const pressedHandle = () => {
    onCancelSuccessfully();
  };
  return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoCancelInvitationButton, {
    icon: require('../resources/icon_request_cohost.png'),
    backgroundColor: 'rgba(30, 39, 64, 0.4)',
    width: 195,
    height: 36,
    fontSize: 13,
    color: "#fff",
    text: _defines.ZegoTranslationText.cancelRequestCoHostButton,
    verticalLayout: false,
    invitees: [hostID],
    onPressed: pressedHandle
  });
}
