// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoRequestCoHostButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoRequestCoHostButton;
var _react = _interopRequireDefault(require("react"));
var _defines = require("../services/defines");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoRequestCoHostButton(props) {
  const {
    hostID,
    liveStatus,
    onRequestSuccessfully,
    setIsToastVisable,
    setToastExtendedData
  } = props;

  // Verify whether invitations can be sent
  const willPressedHandle = () => {
    let result = true;
    if (!hostID) {
      setIsToastVisable(true);
      setToastExtendedData({
        type: _defines.ZegoToastType.error,
        text: _defines.ZegoTranslationText.requestCoHostFailed
      });
      result = false;
    } else if (liveStatus !== _defines.ZegoLiveStatus.start) {
      setIsToastVisable(true);
      setToastExtendedData({
        type: _defines.ZegoToastType.error,
        text: _defines.ZegoTranslationText.requestCoHostFailed
      });
      result = false;
    }
    return result;
  };
  const pressedHandle = () => {
    setIsToastVisable(true);
    setToastExtendedData({
      type: _defines.ZegoToastType.success,
      text: _defines.ZegoTranslationText.sendRequestCoHostToast
    });
    onRequestSuccessfully();
  };
  return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoSendInvitationButton, {
    icon: require('../resources/icon_request_cohost.png'),
    backgroundColor: 'rgba(30, 39, 64, 0.4)',
    width: 165,
    height: 36,
    fontSize: 13,
    color: "#fff",
    text: _defines.ZegoTranslationText.requestCoHostButton,
    verticalLayout: false,
    invitees: [hostID],
    type: _defines.ZegoInvitationType.requestCoHost,
    onWillPressed: willPressedHandle,
    onPressed: pressedHandle
  });
}
