// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoTopBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoTopBar;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
var _ZegoMinimizingButton = _interopRequireDefault(require("./ZegoMinimizingButton"));
var _ZegoMenuBarButtonName = _interopRequireDefault(require("./ZegoMenuBarButtonName"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoTopBar(props) {
  const {
    menuBarButtons = [],
    onLeave,
    onLeaveConfirmation
  } = props;
  const getButtonByButtonIndex = buttonIndex => {
    switch (buttonIndex) {
      case _ZegoMenuBarButtonName.default.minimizingButton:
        return /*#__PURE__*/_react.default.createElement(_ZegoMinimizingButton.default, null);
      case _ZegoMenuBarButtonName.default.leaveButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoLeaveButton, {
          onLeaveConfirmation: onLeaveConfirmation,
          onPressed: onLeave,
          iconLeave: require('../resources/white_top_button_close.png')
        });
    }
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.topRightBar
  }, menuBarButtons.map(buttonIndex => /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    key: buttonIndex,
    style: styles.customIconContainer
  }, getButtonByButtonIndex(buttonIndex))));
}
const styles = _reactNative.StyleSheet.create({
  topRightBar: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  customIconContainer: {
    marginLeft: 0
  }
});
