// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoEndCoHostButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoEndCoHostButton;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _zegoUikitRn = _interopRequireDefault(require("@zegocloud/zego-uikit-rn"));
var _defines = require("../services/defines");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoEndCoHostButton(props) {
  const {
    onEndSuccessfully,
    setIsDialogVisable,
    setDialogExtendedData
  } = props;
  const pressedHandle = () => {
    setIsDialogVisable(true);
    setDialogExtendedData({
      title: _defines.ZegoTranslationText.endConnectionDialogInfo.title,
      content: _defines.ZegoTranslationText.endConnectionDialogInfo.message,
      cancelText: _defines.ZegoTranslationText.endConnectionDialogInfo.cancelButtonName,
      okText: _defines.ZegoTranslationText.endConnectionDialogInfo.confirmButtonName,
      onCancel: () => {
        setIsDialogVisable(false);
      },
      onOk: () => {
        _zegoUikitRn.default.turnCameraOn('', false);
        _zegoUikitRn.default.turnMicrophoneOn('', false);
        onEndSuccessfully();
        setIsDialogVisable(false);
      }
    });
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    style: styles.btnContainer,
    onPress: pressedHandle
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    style: styles.image,
    resizeMode: "contain",
    source: require('../resources/icon_request_cohost.png')
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, _defines.ZegoTranslationText.endCoHostButton));
}
const styles = _reactNative.StyleSheet.create({
  btnContainer: {
    width: 84,
    height: 36,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(255, 13, 35, 0.6)',
    borderRadius: 1000
  },
  image: {
    marginRight: 6
  },
  text: {
    color: '#fff'
  }
});
