// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/commonjs/components/ZegoToast.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoToast;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _defines = require("../services/defines");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ZegoToast(props) {
  const {
    visable = false,
    text = '',
    type = _defines.ZegoToastType.success
  } = props;
  let color = '#FFFFFF',
    backgroundColor = '#55BC9E';
  switch (type) {
    case _defines.ZegoToastType.default:
    case _defines.ZegoToastType.info:
      break;
    case _defines.ZegoToastType.success:
      break;
    case _defines.ZegoToastType.warning:
      backgroundColor = '#e6a23c';
      break;
    case _defines.ZegoToastType.error:
      backgroundColor = '#BD5454';
      break;
    default:
      break;
  }
  const getCustomTextStyle = color => _reactNative.StyleSheet.create({
    text: {
      color
    }
  });
  const getCustomContainerStyle = (visable, backgroundColor) => _reactNative.StyleSheet.create({
    customContainer: {
      display: visable ? 'flex' : 'none',
      backgroundColor
    }
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, getCustomContainerStyle(visable, backgroundColor).customContainer]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, getCustomTextStyle(color).text]
  }, text));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    zIndex: 12,
    height: 70,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    top: 40
  },
  text: {
    fontSize: 14
  }
});
