// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/module/components/ZegoDialog.js
import React, { useState, useEffect } from 'react';
import { View, StyleSheet, Text, TouchableOpacity } from 'react-native';
import PrebuiltHelper from "../services/prebuilt_helper";
export default function ZegoDialog(props) {
  const {
    title = '',
    content = '',
    cancelText = 'cancel',
    okText = 'ok',
    onCancel,
    onOk
  } = props;
  const [visable, setVisable] = useState(false);
  const getCustomContainerStyle = visable => StyleSheet.create({
    customContainer: {
      display: visable ? 'flex' : 'none'
    }
  });
  useEffect(() => {
    const callbackID = 'ZegoDialog' + String(Math.floor(Math.random() * 10000));
    PrebuiltHelper.getInstance().onZegoDialogTrigger(callbackID, visable => {
      setVisable(visable);
    });
    return () => {
      PrebuiltHelper.getInstance().onZegoDialogTrigger(callbackID);
    };
  }, []);
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, getCustomContainerStyle(visable).customContainer]
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.mask
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.main
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.header
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.title
  }, title)), /*#__PURE__*/React.createElement(View, {
    style: styles.content
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.desc
  }, content)), /*#__PURE__*/React.createElement(View, {
    style: styles.footer
  }, /*#__PURE__*/React.createElement(TouchableOpacity, {
    onPress: onCancel,
    style: styles.opaButton
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.opaText
  }, cancelText)), /*#__PURE__*/React.createElement(TouchableOpacity, {
    onPress: onOk,
    style: [styles.opaButton, {
      borderLeftWidth: 1,
      borderLeftColor: 'rgba(255, 255, 255, 0.1)'
    }]
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.opaText
  }, okText))))));
}
const styles = StyleSheet.create({
  container: {
    zIndex: 13,
    width: '100%',
    height: '100%'
  },
  mask: {
    width: '100%',
    height: '100%',
    backgroundColor: 'transparent',
    alignItems: 'center',
    justifyContent: 'center'
  },
  main: {
    width: 300,
    backgroundColor: 'rgba(17, 16, 20, 0.84)',
    borderRadius: 16,
    paddingTop: 15
  },
  header: {
    paddingBottom: 14.5,
    alignItems: 'center',
    justifyContent: 'center',
    paddingLeft: 45,
    paddingRight: 45
  },
  title: {
    color: '#FFFFFF',
    fontSize: 16,
    textAlign: 'center'
  },
  content: {
    paddingBottom: 23,
    alignItems: 'center',
    justifyContent: 'center'
  },
  desc: {
    color: '#FFFFFF',
    fontSize: 14,
    textAlign: 'center',
    paddingLeft: 45,
    paddingRight: 45
  },
  footer: {
    borderTopWidth: 1,
    borderTopColor: 'rgba(255, 255, 255, 0.1)',
    flexDirection: 'row'
  },
  opaButton: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: 14,
    paddingBottom: 13.5
  },
  opaText: {
    color: '#FFFFFF',
    fontSize: 16
  }
});
