// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/module/components/ZegoAudioVideoForegroundView.js
import React from "react";
import { View, Text, StyleSheet } from "react-native";
import { ZegoMicrophoneStateIcon } from '@zegocloud/zego-uikit-rn';
export default function ZegoAudioVideoForegroundView(props) {
  const {
    userInfo,
    showUserNameOnView,
    showCameraStateOnView,
    showMicrophoneStateOnView,
    turnOnMicrophoneWhenJoining,
    turnOnCameraWhenJoining,
    role,
    localUserID
  } = props;
  const {
    userID = '',
    userName = '',
    isMicDeviceOn
  } = userInfo;

  // const panResponder = useRef(PanResponder.create({
  //     onStartShouldSetPanResponderCapture: () => {
  //         console.log(`Toggle the state of other people's devices`);
  //     }
  // })).current;

  return /*#__PURE__*/React.createElement(View, {
    style: styles.foregroundViewContainer
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.bottomContainer
  }, showUserNameOnView ? /*#__PURE__*/React.createElement(View, {
    style: styles.nameLabelContainer
  }, /*#__PURE__*/React.createElement(Text, {
    style: styles.nameLabel
  }, userName)) : /*#__PURE__*/React.createElement(View, null), showMicrophoneStateOnView && !isMicDeviceOn ? /*#__PURE__*/React.createElement(ZegoMicrophoneStateIcon, {
    userID: userID,
    style: styles.deviceIcon
  }) : /*#__PURE__*/React.createElement(View, null)));
}
const styles = StyleSheet.create({
  foregroundViewContainer: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 10
  },
  topContainer: {
    paddingTop: 3,
    paddingRight: 3,
    paddingLeft: 3,
    paddingBottom: 3,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  bottomContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    backgroundColor: '#2A2A2A',
    opacity: 0.5,
    position: 'absolute',
    alignSelf: 'center',
    paddingLeft: 3,
    paddingRight: 3,
    paddingBottom: 1,
    paddingTop: 1,
    borderRadius: 4,
    bottom: 3,
    right: 3
  },
  nameLabelContainer: {
    alignSelf: 'center'
  },
  nameLabel: {
    color: '#FFFFFF',
    fontSize: 10
  },
  deviceIcon: {
    flex: 1,
    position: 'absolute'
  }
});
