// @zegocloud/zego-uikit-prebuilt-live-streaming-rn/lib/module/components/ZegoTopBar.js
import React from 'react';
import { View, StyleSheet } from 'react-native';
import { ZegoLeaveButton } from '@zegocloud/zego-uikit-rn';
import ZegoMinimizingButton from './ZegoMinimizingButton';
import ZegoMenuBarButtonName from "./ZegoMenuBarButtonName";
export default function ZegoTopBar(props) {
  const {
    menuBarButtons = [],
    onLeave,
    onLeaveConfirmation
  } = props;
  const getButtonByButtonIndex = buttonIndex => {
    switch (buttonIndex) {
      case ZegoMenuBarButtonName.minimizingButton:
        return /*#__PURE__*/React.createElement(ZegoMinimizingButton, null);
      case ZegoMenuBarButtonName.leaveButton:
        return /*#__PURE__*/React.createElement(ZegoLeaveButton, {
          onLeaveConfirmation: onLeaveConfirmation,
          onPressed: onLeave,
          iconLeave: require('../resources/white_top_button_close.png')
        });
    }
  };
  return /*#__PURE__*/React.createElement(View, {
    style: styles.topRightBar
  }, menuBarButtons.map(buttonIndex => /*#__PURE__*/React.createElement(View, {
    key: buttonIndex,
    style: styles.customIconContainer
  }, getButtonByButtonIndex(buttonIndex))));
}
const styles = StyleSheet.create({
  topRightBar: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  customIconContainer: {
    marginLeft: 0
  }
});
