// zego-zim-react-native/lib/impl/ZIMLogger.js
import { NativeModules } from 'react-native';
const { ZIMNativeModule } = NativeModules;
export class ZIMLogger {
    warn(tag, action, msg) {
        const log = msg && typeof msg != 'string' ? this.stringify(msg) : msg || '';
        if (log) {
            console.log(action, log);
            tag && ZIMNativeModule.writeCustomLog(log, action);
        }
    }
    /**
     * Format the log to reduce a large number of invalid logs
     *
     * 1. Uint8Array           ->  b=length
     * 2. Array.length > 2     ->  [length, Array[first], Array[last]]
     * 3. String.length > 100  ->  s=length
     * 4. ZIMMessage.message   ->  s=length
     *
     */
    stringify(obj) {
        let msgObj = null;
        const str = JSON.stringify(obj, (key, value) => {
            if (value instanceof Uint8Array)
                return 'b=' + value.length;
            if (value instanceof Array && value.length > 2) {
                const len = value.length;
                return [len, value[0], value[len - 1]];
            }
            if (value && typeof value == 'object' && value.type && value.message)
                msgObj = value;
            if (typeof value == 'string') {
                if (value.length > 100)
                    return 's=' + value.length;
                if (key == 'message' && msgObj && msgObj.type) {
                    msgObj = null;
                    return 's=' + value.length;
                }
            }
            return value;
        });
        return str.replace(/\"(\w+)\":/g, '$1:');
    }
}
export var ZIMLogTag;
(function (ZIMLogTag) {
    ZIMLogTag["User"] = "User";
    ZIMLogTag["Conversation"] = "Conv";
    ZIMLogTag["Room"] = "Room";
    ZIMLogTag["Group"] = "Group";
    ZIMLogTag["Call"] = "Call";
    ZIMLogTag["Database"] = "DB";
})(ZIMLogTag || (ZIMLogTag = {}));
export var ZIMLogAction;
(function (ZIMLogAction) {
    // API - Main
    ZIMLogAction["Login"] = "JSAPI.login";
    // API - User
    ZIMLogAction["QueryUsersInfo"] = "JSAPI.queryUsersInfo";
    ZIMLogAction["UpdateUserName"] = "JSAPI.updateUserName";
    ZIMLogAction["UpdateUserAvatarUrl"] = "JSAPI.updateUserAvatarUrl";
    ZIMLogAction["UpdateUserExtendedData"] = "JSAPI.updateUserExtendedData";
    // API - Conversation
    ZIMLogAction["QueryConversation"] = "JSAPI.queryConversation";
    ZIMLogAction["QueryConversationList"] = "JSAPI.queryConversationList";
    ZIMLogAction["QueryPinnedList"] = "JSAPI.queryConversationPinnedList";
    ZIMLogAction["DeleteConversation"] = "JSAPI.deleteConversation";
    ZIMLogAction["ClearUnreadMessageCount"] = "JSAPI.clearConversationUnreadMessageCount";
    ZIMLogAction["SetNotificationStatus"] = "JSAPI.setConversationNotificationStatus";
    ZIMLogAction["UpdatePinnedState"] = "JSAPI.updateConversationPinnedState";
    // API - Message
    ZIMLogAction["SendMessage"] = "JSAPI.sendMessage";
    ZIMLogAction["SendMediaMessage"] = "JSAPI.sendMediaMessage";
    ZIMLogAction["DeleteMessages"] = "JSAPI.deleteMessages";
    ZIMLogAction["DeleteAllMessage"] = "JSAPI.deleteAllMessage";
    ZIMLogAction["QueryHistoryMessage"] = "JSAPI.queryHistoryMessage";
    ZIMLogAction["DownloadMediaFile"] = "JSAPI.downloadMediaFile";
    ZIMLogAction["InsertMessageToLocalDB"] = "JSAPI.insertMessageToLocalDB";
    ZIMLogAction["UpdateMessageLocalExtendedData"] = "JSAPI.updateMessageLocalExtendedData";
    ZIMLogAction["RevokeMessage"] = "JSAPI.RevokeMessage";
    // API - Message receipt
    ZIMLogAction["SendReceiptRead"] = "JSAPI.sendConversationMessageReceiptRead";
    ZIMLogAction["SendMessageReceiptsRead"] = "JSAPI.sendMessageReceiptsRead";
    ZIMLogAction["QueryReceiptsInfo"] = "JSAPI.queryMessageReceiptsInfo";
    ZIMLogAction["QueryReceiptReadMemberList"] = "JSAPI.queryGroupMessageReceiptReadMemberList";
    ZIMLogAction["QueryReceiptUnreadMemberList"] = "JSAPI.queryGroupMessageReceiptUnreadMemberList";
    // API - Message reaction
    ZIMLogAction["AddReaction"] = "JSAPI.addMessageReaction";
    ZIMLogAction["DeleteReaction"] = "JSAPI.deleteMessageReaction";
    ZIMLogAction["QueryReaction"] = "JSAPI.queryMessageReactionUserList";
    // API - Room
    ZIMLogAction["CreateRoom"] = "JSAPI.createRoom";
    ZIMLogAction["JoinRoom"] = "JSAPI.joinRoom";
    ZIMLogAction["LeaveRoom"] = "JSAPI.leaveRoom";
    ZIMLogAction["QueryRoomMemberList"] = "JSAPI.queryRoomMemberList";
    ZIMLogAction["QueryRoomMembers"] = "JSAPI.queryRoomMembers";
    ZIMLogAction["QueryRoomOnlineMemberCount"] = "JSAPI.queryRoomOnlineMemberCount";
    ZIMLogAction["SetRoomAttributes"] = "JSAPI.setRoomAttributes";
    ZIMLogAction["DeleteRoomAttributes"] = "JSAPI.deleteRoomAttributes";
    ZIMLogAction["QueryRoomAllAttributes"] = "JSAPI.queryRoomAllAttributes";
    ZIMLogAction["BeginRoomAttributesBatchOperation"] = "JSAPI.beginRoomAttributesBatchOperation";
    ZIMLogAction["EndRoomAttributesBatchOperation"] = "JSAPI.endRoomAttributesBatchOperation";
    ZIMLogAction["SetRoomMembersAttributes"] = "JSAPI.setRoomMembersAttributes";
    ZIMLogAction["QueryRoomMembersAttributes"] = "JSAPI.queryRoomMembersAttributes";
    ZIMLogAction["QueryRoomMemberAttributesList"] = "JSAPI.queryRoomMemberAttributesList";
    // API - Group
    ZIMLogAction["CreateGroup"] = "JSAPI.createGroup";
    ZIMLogAction["EnterRoom"] = "JSAPI.enterRoom";
    ZIMLogAction["JoinGroup"] = "JSAPI.joinGroup";
    ZIMLogAction["DismissGroup"] = "JSAPI.dismissGroup";
    ZIMLogAction["LeaveGroup"] = "JSAPI.leaveGroup";
    ZIMLogAction["InviteUsersIntoGroup"] = "JSAPI.inviteUsersIntoGroup";
    ZIMLogAction["KickGroupMembers"] = "JSAPI.kickGroupMembers";
    ZIMLogAction["QueryGroupList"] = "JSAPI.queryGroupList";
    ZIMLogAction["QueryGroupMemberList"] = "JSAPI.queryGroupMemberList";
    ZIMLogAction["QueryGroupMemberCount"] = "JSAPI.queryGroupMemberCount";
    ZIMLogAction["TransferGroupOwner"] = "JSAPI.transferGroupOwner";
    ZIMLogAction["QueryGroupInfo"] = "JSAPI.queryGroupInfo";
    ZIMLogAction["UpdateGroupName"] = "JSAPI.updateGroupName";
    ZIMLogAction["UpdateGroupNotice"] = "JSAPI.updateGroupNotice";
    ZIMLogAction["UpdateGroupAvatarUrl"] = "JSAPI.updateGroupAvatarUrl";
    ZIMLogAction["SetGroupAttributes"] = "JSAPI.setGroupAttributes";
    ZIMLogAction["DeleteGroupAttributes"] = "JSAPI.deleteGroupAttributes";
    ZIMLogAction["QueryGroupAttributes"] = "JSAPI.queryGroupAttributes";
    ZIMLogAction["SetGroupMemberNickname"] = "JSAPI.setGroupMemberNickname";
    ZIMLogAction["SetGroupMemberRole"] = "JSAPI.setGroupMemberRole";
    ZIMLogAction["QueryGroupMemberInfo"] = "JSAPI.queryGroupMemberInfo";
    // API - Call
    ZIMLogAction["CallInvite"] = "JSAPI.callInvite";
    ZIMLogAction["CallCancel"] = "JSAPI.callCancel";
    ZIMLogAction["CallAccept"] = "JSAPI.callAccept";
    ZIMLogAction["CallReject"] = "JSAPI.callReject";
    ZIMLogAction["CallQuit"] = "JSAPI.callQuit";
    ZIMLogAction["CallEnd"] = "JSAPI.callEnd";
    ZIMLogAction["CallingInvite"] = "JSAPI.callingInvite";
    ZIMLogAction["QueryCallList"] = "JSAPI.queryCallInvitationList";
    // API - DB Search
    ZIMLogAction["SearchConversations"] = "JSAPI.searchLocalConversations";
    ZIMLogAction["SearchGlobalMessages"] = "JSAPI.searchGlobalLocalMessages";
    ZIMLogAction["SearchMessages"] = "JSAPI.searchLocalMessages";
    ZIMLogAction["SearchGroups"] = "JSAPI.searchLocalGroups";
    ZIMLogAction["SearchGroupMembers"] = "JSAPI.searchLocalGroupMembers";
})(ZIMLogAction || (ZIMLogAction = {}));
