// zego-zpns-react-native/lib/CallKitDefines.js
export var CXHandleType;
(function (CXHandleType) {
    CXHandleType[CXHandleType["Generic"] = 1] = "Generic";
    CXHandleType[CXHandleType["PhoneNumber"] = 2] = "PhoneNumber";
    CXHandleType[CXHandleType["EmailAddress"] = 3] = "EmailAddress";
})(CXHandleType || (CXHandleType = {}));
export var CXCallEndedReason;
(function (CXCallEndedReason) {
    /// An error occurred while trying to service the call
    CXCallEndedReason[CXCallEndedReason["Failed"] = 1] = "Failed";
    /// The remote party explicitly ended the call
    CXCallEndedReason[CXCallEndedReason["RemoteEnded"] = 2] = "RemoteEnded";
    /// The call never started connecting and was never explicitly ended (e.g. outgoing/incoming call timeout)
    CXCallEndedReason[CXCallEndedReason["Unanswered"] = 3] = "Unanswered";
    /// The call was answered on another device
    CXCallEndedReason[CXCallEndedReason["AnsweredElsewhere"] = 4] = "AnsweredElsewhere";
    /// The call was declined on another device
    CXCallEndedReason[CXCallEndedReason["DeclinedElsewhere"] = 5] = "DeclinedElsewhere";
})(CXCallEndedReason || (CXCallEndedReason = {}));
